### ISSUE TITLE
how to train custom model using deepsomatic

### ISSUE THREAD

**Issue by LiuKe98** [google/deepsomatic#60]
Hello everyone,
   I'm new to deepsomatic and deep learning. I‘m trying to train a custom model using my own dataset with deepsomatic, but I couldn't find any training instructions in the documentation. I noticed there's a `deepvariant-training-case-study.md` file, but it seems to be different from deepsomatic. Also, I didn't see any
train script in the deepsomatic 
Could someone please guide me on how to train a model with DeepSomatic, or point me to any relevant documentation or example scripts?

Thank you very much!

**Comment by danielecook** [comment]
@LiuKe98 we do not have a training case study for deepsomatic currently.  If you have (a) a well-labeled training dataset and (b) a plan for evaluating trained models, than I can provide some guidance on how to get started.

**Comment by LiuKe98** [comment]
Dear Daniel,
    Thank you very much for your quick and helpful response!
    Based on the deepvariant-training-case-study.md, I've attempted to adapt a training pipeline for DeepSomatic using my own dataset. I’ve attached the script for your reference. In the script, I used make_examples_somatic from Deepsomatic in training mode for training and validation sets, and used train.zip from DeepVariant-1.8.0 to train a custom model. 
    I would be truly grateful if you could kindly take a look and let me know whether the approach seems reasonable.
Best regards
[ds.training.txt](https://github.com/user-attachments/files/19847752/ds.training.txt)

**Comment by danielecook** [comment]
@LiuKe98 this looks like a great start. Because we have not released a training case study for DeepSomatic, I cannot promise that training will work for you, but you are welcome to try. 

In looking at your script, there are a few additional flags you will need:

```bash
--customized_classes_labeler_classes_list=ref,germline,somatic \
--customized_classes_labeler_info_field_name=type \
--labeler_algorithm=customized_classes_labeler \
--sample_name_normal=normal \
--sample_name_tumor=tumor
```

Your truth VCF should be formatted like this, where the info field specifies 'TYPE'

```
##INFO=<ID=type,Number=1,Type=String,Description="Type: germline or somatic">
...
chr1    34632968        .       G       A       13.7    PASS    type=germline
chr1    34635370        .       T       C       47.8    PASS    type=germline
chr1    34637674        .       T       C       54.3    PASS    type=germline
chr1    34644178        .       C       CT      30      PASS    type=germline
chr1    34647943        .       C       T       .       PASS    type=somatic
chr1    34652430        .       C       T       63      PASS    type=germline
chr1    34652734        .       C       T       64      PASS    type=germline
chr1    34653989        .       TCA     T       46.4    PASS    type=germline
chr1    34654421        .       G       A       45.3    PASS    type=germline
chr1    34659850        .       C       T       46.4    PASS    type=germline
```

**Comment by LiuKe98** [comment]
Hi  Daniel,
    I’m working on training a custom model using Deepsomatic, and I have two questions I would greatly appreciate your help with.
      **1.** I want to use BAM files from multiple samples for training. The script _ds.training.multisam.getopt.sh.txt_ has been attached. I would be grateful if you could take a look and let me know whether the pipeline seems reasonable. In the script, I first run **make_examples_somatic** from Deepsomatic in training mode for both the training and validation sets, **running one sample at a time**. After that, I **shuffle** all TFRecord files together. Finally, I use **train.zip** from DeepVariant-1.8.0 to train a custom model based on  training_set.dataset_config.pbtxt and validation_set.dataset_config.pbtxt.
      **2.** During the training process, I noticed that in the _training_set.dataset_config.pbtxt_ I attached, only one variant is labeled as class2. However, in the truth VCF file, there are several somatic variants. In the output file _logs.training.twosamid.txt_ I also attached, there are no somatic sites showing "no good match found" or other error information. Is it more likely that **run_examples** didn’t recognize these somatic candidate variants?

I would be truly grateful for any guidance you can provide.

[ds.training.multisam.getopt.sh.txt](https://github.com/user-attachments/files/19957190/ds.training.multisam.getopt.sh.txt)

[training_set.dataset_config.pbtxt.txt](https://github.com/user-attachments/files/19957192/training_set.dataset_config.pbtxt.txt)

[logs.training.twosamid.txt](https://github.com/user-attachments/files/19957193/logs.training.twosamid.txt)

**Comment by akolesnikov** [comment]
* `ds.training.multisam.getopt.sh.txt` script looks right to me. 
* How many class 2 examples are expected? Could you manually check that there are matching alleles in tumor samples? There are multiple reasons the labeler cannot match candidates including: candidates are outside the confident region, too few reads support the matching allele, same allele present in normal sample, supporting reads are downsampled if total number of reads per region exceeds 1500.

**Comment by chstubbe** [comment]
Hi!

I am trying to train deepvariant on RNA SEQ TUMOR ONLY SOMATIC variant calling.
I was already in contact with @danielecook regarding some guidelines on this. 

I have prepared my input data:
- vcfs in right output format (these vcfs are filtered on coverage > 10 based on the RNA bam file)
- RNA tumor bam (chr split for eval/train/test)
- exons.bed file 
- GRCh38.d1.vd1.fa reference
I use the deepvariant version 1.9.0 sif image. 

<img width="1106" height="219" alt="Image" src="https://github.com/user-attachments/assets/7a21f78c-60f1-4828-afa8-ecc5a9ba70ee" />

Is the header of my vcf file too extensive? 

Then, I ran make_examples for train/test/eval and created the tfrecords files. 
Below are the specific parameters I used:
    --mode training \
    --customized_classes_labeler_classes_list=ref,germline,somatic \
    --customized_classes_labeler_info_field_name=type \
    --split_skip_reads=True \
    --channel_list='BASE_CHANNELS,insert_size'
I then shuffled these records, and obtained pbtxt files. 

But when I looked deeper in my records files, I see that all the variants get label 0 (ref). 
For example for one training sample and first 10000 examples: 
Variants in truth vcf: 5609 germline, 73 somatic
Class label distribution:
  ref (class0):   10,000  (100.0%)
                 TOTAL:   10,000
I also checked the overlap with the exons.bed file and this was: 5636, so this should not be the problem I think. 

Is there someone that would have any insight on this? Some help would be greatly appreciated!

With kind regards
Charlotte
